### PR TITLE
Set explain attributes after the query is prepared

### PR DESCRIPTION
**Pull request by abhishekrb19**
In this PR:
- Set the explain plan attributes _after_ the query is prepared,  when we've the finalized output names in `rootQueryRel`. This also gives us the benefit of going through all validation checks in the prepare phase.
- Simplify `resolveClusteredByColumnsToOutputColumns` function for explain attributes as things are already validated and we just have one source list, `rootQueryRel`, containing the output column names; the existing unit tests work right out of the box. Added new unit tests using `JOIN` and `WITH...AS` to prove that the logic no longer depends on individual `SqlNode` types.
- Add a negative ordinal check for clustered by columns as a workaround to Calcite's limitation; added unit tests to validate this.

##### Key changed/added classes in this PR
 * `DruidSqlParserUtils.java`
 * `QueryHandler.java`
 * `DruidSqlParserUtilsTest.java`
 * `CalciteInsertDmlTest.java`

This PR has:

- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x]  added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by abhishekrb19** [comment]
@gianm, thanks for the review. Yes, I went ahead with 1 -- moving this logic to _after_ the query is _prepared_ enables us to extract the necessary output columns from `rootQueryRel` always - instead of relying on the different `SqlNode` types in the pre-prepare phase. Another benefit of doing it after the query preparation phase is that it already goes through the validation checks. IMO, the code looks simpler.

File: sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java [hunk 1]
```diff
@@ -327,84 +325,54 @@ public static SqlOrderBy convertClusterByToOrderBy(SqlNode query, SqlNodeList cl
...
-              "Node must be a SqlIdentifier, but found [%s]",
-              node.getKind()
-          );
-          SqlIdentifier n = ((SqlIdentifier) node);
-          retClusteredByNames.add(n.isSimple() ? n.getSimple() : n.names.get(1));
-        }
+      if (clusteredByNode instanceof SqlNumericLiteral) {
+        // The node is a literal number -- an ordinal in the CLUSTERED BY clause. Lookup the ordinal in field mappings.
+        final int ordinal = ((SqlNumericLiteral) clusteredByNode).getValueAs(Integer.class);
+        retClusteredByNames.add(sourceFieldMappings.get(ordinal - 1).right);
```
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:357]: are the bounds of the ordinal validated elswhere? I see we are getting rid of that explicit validation that was previously done here?

Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:357]: nvm see comment below.

File: sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java [hunk 2]
```diff
@@ -327,84 +325,54 @@ public static SqlOrderBy convertClusterByToOrderBy(SqlNode query, SqlNodeList cl
...
       } else if (clusteredByNode instanceof SqlBasicCall) {
         // The node is an expression/operator.
-        retClusteredByNames.add(getColumnNameFromSqlCall(clusteredByNode));
+        retClusteredByNames.add(getColumnNameFromSqlCall((SqlBasicCall) clusteredByNode));
       } else {
```
Comment (zachjsh) [inline sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:361]: Does this else clause still assume a certain kind of sqlNode? If so, should we make `else if` and the last `else` branch will log a warning / error about unexpected sqlNode type, but still continue?

File: sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java (hunk 1 above)
Comment (abhishekrb19) [reply sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:357]: yes, Calcite throws an ordinal out of range exception - I also added a negative ordinal check, which later Calcite versions seems to fix. As described in the PR summary, with the attribute setting done _after_ the query is prepared, we leverage all the validations that Calcite does in addition to semantic checking we do in the SQL planner. Please see the unit tests `testReplaceWithNonExistentOrdinalInClusteredBy` and `testReplaceWithNegativeOrdinalInClusteredBy` that demonstrate that invalid ordinals are handled gracefully

File: sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java (hunk 2 above)
Comment (abhishekrb19) [reply sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:361]: This block is a catch-all and doesn't assume any specific node type. I don't think we need to do any special handling to get the output columns other than ordinals and expressions, so this simple node `toString()` should suffice. Also, the rationale is that if we need to constrain the node type for the `CLUSTERED BY` grammar or add validation checks, that needs to be in the SQL planner code outside this explain plan logic.

File: sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java (hunk 1 above)
Comment (zachjsh) [reply sql/src/main/java/org/apache/druid/sql/calcite/parser/DruidSqlParserUtils.java:357]: sounds good!

**Review by zachjsh — APPROVED** [review]
🚀

**Pull request merged**